File: app/src/main/java/tn/esprit/wayfinder/ui/components/ReviewsSection.kt
package tn.esprit.wayfinder.ui.components

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import coil.compose.AsyncImage
import tn.esprit.wayfinder.R
import tn.esprit.wayfinder.models.Review
import tn.esprit.wayfinder.models.ReviewStatsResponse
import tn.esprit.wayfinder.viewmodels.ReviewsUiState
import tn.esprit.wayfinder.viewmodels.ReviewsViewModel
import java.text.SimpleDateFormat
import java.util.*

@Composable
fun ReviewsSection(
    itemType: String,
    itemId: String,
    reviewsViewModel: ReviewsViewModel
) {
    val uiState by reviewsViewModel.uiState.collectAsState()
    var showReviewDialog by remember { mutableStateOf(false) }

    LaunchedEffect(itemId) {
        reviewsViewModel.loadReviews(itemType, itemId)
    }

    Card(
        modifier = Modifier.fillMaxWidth(),
        shape = RoundedCornerShape(16.dp),
        elevation = CardDefaults.cardElevation(4.dp)
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(20.dp),
            verticalArrangement = Arrangement.spacedBy(16.dp)
        ) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically
            ) {
                Text(
                    text = "Avis",
                    fontSize = 20.sp,
                    fontWeight = FontWeight.Bold
                )
                TextButton(onClick = { showReviewDialog = true }) {
                    Text("Ajouter un avis")
                }
            }

            when (val state = uiState) {
                is ReviewsUiState.Loading -> {
                    Box(
                        modifier = Modifier
                            .fillMaxWidth()
                            .height(100.dp),
                        contentAlignment = Alignment.Center
                    ) {
                        CircularProgressIndicator()
                    }
                }
                is ReviewsUiState.Success -> {
                    state.stats?.let { stats ->
                        ReviewStatsDisplay(stats = stats)
                    }

                    if (state.reviews.isEmpty()) {
                        Text(
                            text = "Aucun avis pour le moment. Soyez le premier à laisser un avis!",
                            color = Color.Gray,
                            modifier = Modifier.padding(vertical = 16.dp)
                        )
                    } else {
                        LazyColumn(
                            modifier = Modifier.height(400.dp),
                            verticalArrangement = Arrangement.spacedBy(12.dp)
                        ) {
                            items(state.reviews) { review ->
                                ReviewCard(
                                    review = review,
                                    isUserReview = state.userReview?.id == review.id,
                                    onEdit = { showReviewDialog = true },
                                    onDelete = {
                                        reviewsViewModel.deleteReview(review.id, itemType, itemId)
                                    }
                                )
                            }
                        }
                    }
                }
                is ReviewsUiState.Error -> {
                    Text(
                        text = state.message,
                        color = Color.Red,
                        modifier = Modifier.padding(vertical = 16.dp)
                    )
                }
                else -> {}
            }
        }
    }

    if (showReviewDialog) {
        ReviewDialog(
            itemType = itemType,
            itemId = itemId,
            existingReview = (uiState as? ReviewsUiState.Success)?.userReview,
            onDismiss = { showReviewDialog = false },
            onSubmit = { rating, comment ->
                val existingReview = (uiState as? ReviewsUiState.Success)?.userReview
                if (existingReview != null) {
                    reviewsViewModel.updateReview(
                        existingReview.id,
                        itemType,
                        itemId,
                        rating,
                        comment,
                        null
                    )
                } else {
                    reviewsViewModel.createReview(itemType, itemId, rating, comment, null)
                }
                showReviewDialog = false
            }
        )
    }
}

@Composable
fun ReviewStatsDisplay(stats: ReviewStatsResponse) {
    Column(
        modifier = Modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically
        ) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Text(
                    text = String.format("%.1f", stats.averageRating),
                    fontSize = 32.sp,
                    fontWeight = FontWeight.Bold,
                    color = Color(0xFF1976D2)
                )
                Spacer(modifier = Modifier.width(8.dp))
                StarRating(
                    rating = stats.averageRating.toInt(),
                    modifier = Modifier.size(20.dp)
                )
            }
            Text(
                text = "${stats.totalReviews} avis",
                fontSize = 14.sp,
                color = Color.Gray
            )
        }

        // Rating distribution
        Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
            (5 downTo 1).forEach { stars ->
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Text(
                        text = "$stars",
                        fontSize = 12.sp,
                        modifier = Modifier.width(16.dp)
                    )
                    Icon(
                        imageVector = Icons.Filled.Star,
                        contentDescription = null,
                        tint = Color(0xFFFFD700),
                        modifier = Modifier.size(12.dp)
                    )
                    Spacer(modifier = Modifier.width(8.dp))
                    LinearProgressIndicator(
                        progress = {
                            val count = stats.ratingDistribution[stars.toString()] ?: 0
                            if (stats.totalReviews > 0) count.toFloat() / stats.totalReviews else 0f
                        },
                        modifier = Modifier
                            .weight(1f)
                            .height(8.dp),
                        color = Color(0xFFFFD700),
                        trackColor = Color.LightGray
                    )
                    Spacer(modifier = Modifier.width(8.dp))
                    Text(
                        text = "${stats.ratingDistribution[stars.toString()] ?: 0}",
                        fontSize = 12.sp,
                        color = Color.Gray,
                        modifier = Modifier.width(24.dp)
                    )
                }
            }
        }
    }
}

@Composable
fun ReviewCard(
    review: Review,
    isUserReview: Boolean,
    onEdit: () -> Unit,
    onDelete: () -> Unit
) {
    Card(
        modifier = Modifier.fillMaxWidth(),
        shape = RoundedCornerShape(12.dp),
        colors = CardDefaults.cardColors(containerColor = MaterialTheme.colorScheme.surface),
        elevation = CardDefaults.cardElevation(2.dp)
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(12.dp),
            verticalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically
            ) {
                Row(verticalAlignment = Alignment.CenterVertically) {
                    val profileImageUrl = review.userId.profileImageUrl?.let { url ->
                        if (url.startsWith("http")) url else "https://wayfinder-api-w92x.onrender.com$url"
                    } ?: "https://i.pravatar.cc/150?img=${kotlin.math.abs(review.userId.id.hashCode()) % 70}"
                    AsyncImage(
                        model = profileImageUrl,
                        contentDescription = "User Avatar",
                        modifier = Modifier
                            .size(32.dp)
                            .clip(CircleShape),
                        contentScale = ContentScale.Crop,
                        placeholder = painterResource(id = R.drawable.europe),
                        error = painterResource(id = R.drawable.europe)
                    )
                    Spacer(modifier = Modifier.width(8.dp))
                    Column {
                        Text(
                            text = buildString {
                                val firstName = review.userId.firstName?.takeIf { it.isNotBlank() }
                                val lastName = review.userId.lastName?.takeIf { it.isNotBlank() }
                                when {
                                    firstName != null && lastName != null -> append("$firstName $lastName")
                                    firstName != null -> append(firstName)
                                    lastName != null -> append(lastName)
                                    else -> append(review.userId.username)
                                }
                            },
                            fontSize = 14.sp,
                            fontWeight = FontWeight.Bold
                        )
                        StarRating(
                            rating = review.rating,
                            modifier = Modifier.size(12.dp)
                        )
                    }
                }
                if (isUserReview) {
                    Row {
                        IconButton(onClick = onEdit, modifier = Modifier.size(24.dp)) {
                            Icon(
                                Icons.Filled.Edit,
                                contentDescription = "Edit",
                                modifier = Modifier.size(16.dp),
                                tint = Color(0xFF1976D2)
                            )
                        }
                        IconButton(onClick = onDelete, modifier = Modifier.size(24.dp)) {
                            Icon(
                                Icons.Filled.Delete,
                                contentDescription = "Delete",
                                modifier = Modifier.size(16.dp),
                                tint = Color.Red
                            )
                        }
                    }
                }
            }
            review.comment?.let {
                Text(
                    text = it,
                    fontSize = 14.sp,
                    color = Color.Gray
                )
            }
            Text(
                text = formatTimestamp(review.createdAt),
                fontSize = 12.sp,
                color = Color.Gray
            )
        }
    }
}

@Composable
fun StarRating(
    rating: Int,
    modifier: Modifier = Modifier
) {
    Row(modifier = modifier) {
        repeat(5) { index ->
            Icon(
                imageVector = Icons.Filled.Star,
                contentDescription = null,
                tint = if (index < rating) Color(0xFFFFD700) else Color.LightGray,
                modifier = Modifier.size(12.dp)
            )
        }
    }
}

@Composable
fun ReviewDialog(
    itemType: String,
    itemId: String,
    existingReview: Review?,
    onDismiss: () -> Unit,
    onSubmit: (Int, String?) -> Unit
) {
    var rating by remember { mutableStateOf(existingReview?.rating ?: 0) }
    var comment by remember { mutableStateOf(existingReview?.comment ?: "") }

    AlertDialog(
        onDismissRequest = onDismiss,
        title = { Text(if (existingReview != null) "Modifier votre avis" else "Ajouter un avis") },
        text = {
            Column(verticalArrangement = Arrangement.spacedBy(16.dp)) {
                Text("Note")
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.Center
                ) {
                    (1..5).forEach { star ->
                        IconButton(onClick = { rating = star }) {
                            Icon(
                                imageVector = Icons.Filled.Star,
                                contentDescription = "$star stars",
                                tint = if (star <= rating) Color(0xFFFFD700) else Color.LightGray,
                                modifier = Modifier.size(32.dp)
                            )
                        }
                    }
                }
                OutlinedTextField(
                    value = comment,
                    onValueChange = { comment = it },
                    label = { Text("Commentaire (optionnel)") },
                    modifier = Modifier.fillMaxWidth(),
                    maxLines = 4
                )
            }
        },
        confirmButton = {
            Button(
                onClick = { onSubmit(rating, comment.takeIf { it.isNotBlank() }) },
                enabled = rating > 0
            ) {
                Text(if (existingReview != null) "Modifier" else "Publier")
            }
        },
        dismissButton = {
            TextButton(onClick = onDismiss) {
                Text("Annuler")
            }
        }
    )
}

private fun formatTimestamp(timestamp: String): String {
    return try {
        val parser = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.getDefault())
        parser.timeZone = TimeZone.getTimeZone("UTC")
        val date = parser.parse(timestamp)
        val formatter = SimpleDateFormat("dd MMM yyyy", Locale.getDefault())
        formatter.format(date ?: Date())
    } catch (e: Exception) {
        timestamp
    }
}

File: WayFinder/Views/Home/ReviewsSection.swift
import SwiftUI

struct ReviewsSection: View {
    @Environment(\.colorScheme) private var colorScheme
    @StateObject private var viewModel = ReviewViewModel()
    let itemType: String
    let itemId: String
    @State private var showReviewDialog = false
    @State private var selectedReview: Review?
    
    var body: some View {
        VStack(alignment: .leading, spacing: 16) {
            HStack {
                Text("Avis")
                    .font(.system(size: 20, weight: .bold))
                    .foregroundStyle(ThemeColors.primaryText(colorScheme))
                Spacer()
                Button("Ajouter un avis") {
                    selectedReview = nil
                    showReviewDialog = true
                }
                .font(.subheadline)
                .foregroundColor(Color(red: 0.098, green: 0.463, blue: 0.824))
            }
            
            if viewModel.isLoading {
                ProgressView()
                    .frame(maxWidth: .infinity)
                    .frame(height: 100)
            } else if let errorMessage = viewModel.errorMessage {
                Text(errorMessage)
                    .foregroundColor(.red)
                    .padding(.vertical, 16)
            } else {
                if let stats = viewModel.reviewStats {
                    ReviewStatsDisplay(stats: stats)
                }
                
                if viewModel.reviews.isEmpty {
                    Text("Aucun avis pour le moment. Soyez le premier à laisser un avis!")
                        .foregroundStyle(ThemeColors.secondaryText(colorScheme))
                        .padding(.vertical, 16)
                } else {
                    ScrollView {
                        VStack(spacing: 12) {
                            ForEach(viewModel.reviews) { review in
                                ReviewCard(review: review)
                            }
                        }
                    }
                    .frame(height: 400)
                }
            }
        }
        .padding(20)
        .background(
            RoundedRectangle(cornerRadius: 16)
                .fill(Color.white)
        )
        .shadow(color: Color.black.opacity(0.04), radius: 4, x: 0, y: 4)
        .task {
            await loadReviews()
        }
        .sheet(isPresented: $showReviewDialog) {
            ReviewDialog(
                itemType: itemType,
                itemId: itemId,
                existingReview: selectedReview,
                onDismiss: { showReviewDialog = false },
                onSubmit: { rating, comment in
                    Task {
                        guard let itemTypeEnum = ReviewItemType(rawValue: itemType) else {
                            print("Invalid item type: \(itemType)")
                            showReviewDialog = false
                            return
                        }
                        
                        await viewModel.createReview(
                            itemType: itemTypeEnum,
                            itemId: itemId,
                            rating: rating,
                            comment: comment
                        )
                        await loadReviews()
                        showReviewDialog = false
                    }
                }
            )
        }
    }
    
    private func loadReviews() async {
        // Convert String to ReviewItemType
        guard let itemTypeEnum = ReviewItemType(rawValue: itemType) else {
            print("Invalid item type: \(itemType)")
            return
        }
        await viewModel.loadReviews(itemType: itemTypeEnum, itemId: itemId)
        await viewModel.loadReviewStats(itemType: itemTypeEnum, itemId: itemId)
    }
}

struct ReviewStatsDisplay: View {
    let stats: ReviewStats
    
    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            HStack {
                Text(String(format: "%.1f", stats.averageRating))
                    .font(.system(size: 32, weight: .bold))
                    .foregroundColor(Color(red: 0.098, green: 0.463, blue: 0.824))
                
                StarRating(rating: Int(stats.averageRating.rounded()))
                    .frame(height: 20)
                
                Spacer()
                
                Text("\(stats.totalReviews) avis")
                    .font(.caption)
                    .foregroundStyle(ThemeColors.secondaryText(.light))
            }
            
            // Rating distribution
            if let distribution = stats.ratingDistribution {
                VStack(spacing: 4) {
                    ForEach((1...5).reversed(), id: \.self) { stars in
                        HStack {
                            Text("\(stars)")
                                .font(.caption)
                                .frame(width: 16)
                            
                            Image(systemName: "star.fill")
                                .foregroundColor(Color(red: 1.0, green: 0.843, blue: 0.0))
                                .font(.system(size: 12))
                            
                            GeometryReader { geometry in
                                ZStack(alignment: .leading) {
                                    Rectangle()
                                        .fill(Color.gray.opacity(0.2))
                                        .frame(height: 8)
                                    
                                    Rectangle()
                                        .fill(Color(red: 1.0, green: 0.843, blue: 0.0))
                                        .frame(
                                            width: stats.totalReviews > 0 
                                                ? geometry.size.width * (Double(distribution[stars] ?? 0) / Double(stats.totalReviews))
                                                : 0,
                                            height: 8
                                        )
                                }
                            }
                            .frame(height: 8)
                            
                            Text("\(distribution[stars] ?? 0)")
                                .font(.caption)
                                .foregroundStyle(ThemeColors.secondaryText(.light))
                                .frame(width: 24)
                        }
                    }
                }
            }
        }
    }
}

struct ReviewCard: View {
    @Environment(\.colorScheme) private var colorScheme
    let review: Review
    
    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            HStack {
                Circle()
                    .fill(Color.gray.opacity(0.3))
                    .frame(width: 32, height: 32)
                
                VStack(alignment: .leading, spacing: 4) {
                    Text("Utilisateur")
                        .font(.system(size: 14, weight: .bold))
                    
                    StarRating(rating: review.rating)
                        .frame(height: 12)
                }
                
                Spacer()
            }
            
            if let comment = review.comment, !comment.isEmpty {
                Text(comment)
                    .font(.system(size: 14))
                    .foregroundStyle(ThemeColors.secondaryText(colorScheme))
            }
            
            if let createdAt = review.createdAt {
                Text(formatTimestamp(createdAt))
                    .font(.caption)
                    .foregroundStyle(ThemeColors.secondaryText(colorScheme))
            }
        }
        .padding(12)
        .background(
            RoundedRectangle(cornerRadius: 12)
                .fill(Color.white)
        )
        .shadow(color: Color.black.opacity(0.02), radius: 2, x: 0, y: 2)
    }
    
    private func formatTimestamp(_ date: Date) -> String {
        let displayFormatter = DateFormatter()
        displayFormatter.dateFormat = "dd MMM yyyy"
        return displayFormatter.string(from: date)
    }
}

struct StarRating: View {
    let rating: Int
    var body: some View {
        HStack(spacing: 2) {
            ForEach(0..<5) { index in
                Image(systemName: index < rating ? "star.fill" : "star")
                    .foregroundColor(index < rating ? Color(red: 1.0, green: 0.843, blue: 0.0) : Color.gray.opacity(0.3))
                    .font(.system(size: 12))
            }
        }
    }
}

struct ReviewDialog: View {
    let itemType: String
    let itemId: String
    let existingReview: Review?
    let onDismiss: () -> Void
    let onSubmit: (Int, String?) -> Void
    
    @State private var rating: Int
    @State private var comment: String
    
    init(itemType: String, itemId: String, existingReview: Review?, onDismiss: @escaping () -> Void, onSubmit: @escaping (Int, String?) -> Void) {
        self.itemType = itemType
        self.itemId = itemId
        self.existingReview = existingReview
        self.onDismiss = onDismiss
        self.onSubmit = onSubmit
        _rating = State(initialValue: existingReview?.rating ?? 0)
        _comment = State(initialValue: existingReview?.comment ?? "")
    }
    
    var body: some View {
        NavigationStack {
            VStack(spacing: 24) {
                Text("Note")
                    .font(.headline)
                    .frame(maxWidth: .infinity, alignment: .leading)
                
                HStack(spacing: 16) {
                    ForEach(1...5, id: \.self) { star in
                        Button {
                            rating = star
                        } label: {
                            Image(systemName: star <= rating ? "star.fill" : "star")
                                .font(.system(size: 32))
                                .foregroundColor(star <= rating ? Color(red: 1.0, green: 0.843, blue: 0.0) : Color.gray)
                        }
                    }
                }
                .frame(maxWidth: .infinity)
                
                TextField("Commentaire (optionnel)", text: $comment, axis: .vertical)
                    .textFieldStyle(.roundedBorder)
                    .lineLimit(4...10)
                
                Spacer()
                
                Button {
                    onSubmit(rating, comment.isEmpty ? nil : comment)
                } label: {
                    Text(existingReview != nil ? "Modifier" : "Publier")
                        .font(.headline)
                        .foregroundColor(.white)
                        .frame(maxWidth: .infinity)
                        .padding(.vertical, 14)
                        .background(rating > 0 ? Color(red: 0.098, green: 0.463, blue: 0.824) : Color.gray)
                        .clipShape(RoundedRectangle(cornerRadius: 12))
                }
                .disabled(rating == 0)
            }
            .padding()
            .navigationTitle(existingReview != nil ? "Modifier votre avis" : "Ajouter un avis")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button("Annuler") {
                        onDismiss()
                    }
                }
            }
        }
        .presentationDetents([.medium])
    }
}
